engine/pike2/pike2.go

```diff
@@ -0,0 +1,207 @@
+// Package pike2 provides functionality to work with Pay-to-PubKey-Hash (P2PKH) scripts in a blockchain context.
+//
+// P2PKH is a common type of Bitcoin transaction that locks funds to a specific public key hash, requiring the
+// corresponding private key to produce a valid signature for spending the funds. This package offers utilities
+// to create, decode, and validate P2PKH scripts.
+//
+// The package includes:
+// - Functions to generate P2PKH addresses from public keys.
+// - Methods to construct P2PKH scriptPubKey and scriptSig scripts.
+// - Utilities to decode and inspect P2PKH scripts and addresses.
+// - Validation functions to ensure the integrity and correctness of P2PKH scripts.
+//
+// This package is intended for developers working with Bitcoin or other cryptocurrencies that support P2PKH transactions.
+// It abstracts the low-level details and provides a high-level interface for creating and handling P2PKH scripts.
+package pike2
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/libsv/go-bk/bec"
+	"github.com/libsv/go-bk/crypto"
+	"github.com/libsv/go-bt/v2/bscript"
+	"github.com/libsv/go-bt/v2/bscript/interpreter"
+)
+
+// PikeOutputsTemplate represents the script and satoshis for a Pike output
+type PikeOutputsTemplate struct {
+	Script   string `json:"script"`
+	Satoshis uint64 `json:"satoshis"`
+}
+
+// CreatePikeOutputsTemplate creates a single output with the PIKE template
+func CreatePikeOutputsTemplate(satoshis uint64) ([]PikeOutputsTemplate, error) {
+
+	opcodes := []byte{
+		bscript.OpDUP,
+		bscript.OpHASH160,
+		bscript.OpPUBKEYHASH,
+		bscript.OpEQUALVERIFY,
+		bscript.OpCHECKSIG,
+	}
+
+	// Convert opcodes to hexadecimal string
+	scriptHex := hex.EncodeToString(opcodes)
+
+	// Check if the conversion was successful
+	if scriptHex == "" {
+		return nil, errors.New("failed to create script hex")
+	}
+
+	// Create and return the PikeOutputsTemplate
+	return []PikeOutputsTemplate{
+		{
+			Script:   scriptHex,
+			Satoshis: satoshis,
+		},
+	}, nil
+}
+
+// CalculatePubSharedSecret creates public shared secret base on sender public key
+func CalculatePubSharedSecret(senderPubKey *bec.PublicKey) *bec.PublicKey {
```

Comment:
After some calculations it occurs, that this part (this calculation) makes no sense. Because in output we actually receive exactly the same public key. Sorry it was my misunderstanding of the concept.

Therefore calculating of the shared secret should be remove. And we should just use senderPublicKey instead in the later steps.

Reply:
simply changed to senderPubKey as we were doing multiplying by 1 with private key